### PR TITLE
Add steady-state (SPGR) Bloch simulation support

### PR DESCRIPTION
**Pull request by YongliHe23**
- `SpinArray.applypulse_ss` / `SpinCube.applypulse_ss`: compute the steady-state magnetization before an alpha excitation pulse, assuming a [beta – alpha – readout] x N_rep SPGR sequence structure.
- `utils.target_Mss`: compute target SS magnetization profiles for inner- and outer-volume regions given desired beta flip angles.
- `sims.blochsim`: allow `Beff` batch dim of 1 to broadcast over `Mi` (avoids redundant memory allocation when Beff is the same for all spins).

### CODE REVIEW

File: mrphy/sims.py [hunk 1]
```diff
@@ -302,6 +302,8 @@ def blochsim(
     """
 
     # %% Defaults and move to the same device
+    if Beff.shape[0] != Mi.shape[0]:
+        Beff = Beff.expand(Mi.shape[0], *Beff.shape[1:])
```
Comment (tianrluo) [inline mrphy/sims.py:306]: could you help testing if this explicit `Beff.expand` is necessary?
to my memory, if `Beff` was of shape `(1, *Nd, nT, xyz)`, it would be broadcasted across `N`, and didn't need to expand.
(if broadcasting was the behavior, i indeed missed annotating the ` ⊻ 1` for `Beff` input tho...)

File: mrphy/utils.py [hunk 2]
```diff
@@ -357,3 +359,59 @@ def uϕrot(U: Tensor, Φ: Tensor, Vi: Tensor) -> Tensor:
...
+    d[..., 0] = scale * torch.sin(torch.deg2rad(beta))  # Mxy before alpha (along x)
+    d = d.nan_to_num()
+
+    weight = weight_iv * iv + weight_ov * ov
+    weight = weight[None, ...]
+
+    return {
+        "d": (d if doEmbed else cube.extract(d)),
+        "weight": (weight if doEmbed else cube.extract(weight))
+    }
```
Comment (tianrluo) [inline mrphy/utils.py]: let's not use dictionary as output.
return them just as tuple instead, e.g., `return d, weight`.

`dict` makes it difficult for coding tools to track the types.
i wish i had such habit when i initially wrote the package...
it's more important to make the code human-readable, ide/coding-tool friendly, and, therefore, maintainable.

**Review by tianrluo — CHANGES_REQUESTED** [review]
Besides the comments above, as communicated in our emails,
let's create a new sub-module `mrphy.steady_state` for such applications.

Thanks a lot for adding new features to `MRphy.py`!

File: mrphy/sims.py (hunk 1 above)
Comment (YongliHe23) [reply mrphy/sims.py:306]: This expand is necessary because `BlochSim` object allocates its internal buffers (`Mhist`,etc) using `NNd=Beff.shape[:-2]`, so if `Beff` has batch dim 1, all those buffers are under-allocated and would raise an error later on when the loop tries to write `N`-size results into 1-sized buffers.
Although it seems that in `SpinArray.pulse2beff`, the output `beff` always have batch dim `N` so this might not be an explicit issue. Nevertheless adding this expand should make it more robust. Plus, `expand` does not cost extra memory!

**Comment by YongliHe23** [comment]
Refactored: moved steady-state features into a new `mrphy.steady_state` submodule (`spgr`, `target_spgr`) instead of adding methods directly to `SpinArray/SpinCube`, to keep the spin objects from becoming too cumbersome as more sequence types are added.

File: mrphy/steady_state/spgr.py
Comment (tianrluo) [inline mrphy/steady_state/spgr.py:1]: Let's make `steady_state` a directory rather than a single file, and move the current file into it (after rename).

I am also a bit reluctant to directly call it `spgr`.
There are different ways to do SPGR in general.
A dedicated method of a specialized name can help avoid the necessity to include logic to deal w/ all other sgpr mutants, such logic would very much confound the code maintenance.

you can maybe make the file `mrphy/steady_state/spgr.py`, then name the functions with something like `spgr_ovs`, `spgr_yh`, or other names that are less general.

File: mrphy/steady_state.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import torch
+from torch import Tensor
+
+__all__ = ['spgr', 'target_spgr']
+
+
+def spgr(
+    spin,
+    pulse,
```
Comment (tianrluo) [inline mrphy/steady_state.py]: (I'm not sure, but) you should be able to import `SpinArray, SpinCube, Pulse` classes from `mrphy.mobjs` to the head of the file, and use them for type annotations.

File: mrphy/steady_state.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    *,
+    doEmbed: bool = False, doRelax: bool = True, doUpdate: bool = False,
+    betaoff: bool = False,
+    loc: Optional[Tensor] = None, loc_: Optional[Tensor] = None,
+    Δf: Optional[Tensor] = None, Δf_: Optional[Tensor] = None,
+    b1Map: Optional[Tensor] = None, b1Map_: Optional[Tensor] = None,
+    alpha: float = 0, TR: float = 55e-3
```
Comment (tianrluo) [inline mrphy/steady_state.py]: let's keep arguments in a single-line manner - one argument per line, all ends with `,` for style consistency.
it is easier to read in general.
(another thing i adapted overtime, and wished i started doing it in the beginning...)

i should make a separate commit in the future to change the code style elsewhere in the repo.

File: mrphy/steady_state.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+
+def target_spgr(
+    cube,
+    beta_iv: float, beta_ov: float, iv: Tensor, ov: Tensor,
+    *,
+    weight_iv: float = 1.0, weight_ov: float = 1.0,
+    doEmbed: bool = True,
+    alpha: float = 0, TR: float = 55e-3
+) -> tuple:
```
Comment (tianrluo) [inline mrphy/steady_state.py]: output type annotation could be `tuple[Tensor, Tensor]`.

(sry i should have been more specific in my original comment).

**Review by tianrluo — CHANGES_REQUESTED** [review]

**Comment by YongliHe23** [comment]
I have updated the commit accordingly. Lmk anything else need to be changed!

**Review by tianrluo — APPROVED** [review]
